server/k8s.go

```diff
@@ -110,9 +120,56 @@ func (w *k8sWatcherImpl) startWithLoadedConfig(config *rest.Config) error {
 		},
 	)
 
+	w.mappings = make(map[string]string)
+	_, statefulSetController := cache.NewInformer(
+		cache.NewListWatchFromClient(
+			clientset.AppsV1().RESTClient(),
+			"statefulSets",
+			core.NamespaceAll,
+			fields.Everything(),
+		),
+		&apps.StatefulSet{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				statefulSet, ok := obj.(*apps.StatefulSet)
+				if !ok {
+					return
+				}
+				w.RLock()
+				defer w.RUnlock()
+				w.mappings[statefulSet.Spec.ServiceName] = statefulSet.Name
+			},
+			DeleteFunc: func(obj interface{}) {
+				statefulSet, ok := obj.(*apps.StatefulSet)
+				if !ok {
+					return
+				}
+				w.RLock()
+				defer w.RUnlock()
+				delete(w.mappings, statefulSet.Spec.ServiceName)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				oldStatefulSet, ok := oldObj.(*apps.StatefulSet)
+				if !ok {
+					return
+				}
+				newStatefulSet, ok := newObj.(*apps.StatefulSet)
+				if !ok {
+					return
+				}
+				w.RLock()
+				defer w.RUnlock()
+				delete(w.mappings, oldStatefulSet.Spec.ServiceName)
+				w.mappings[newStatefulSet.Spec.ServiceName] = newStatefulSet.Name
+			},
+		},
+	)
+
 	w.stop = make(chan struct{}, 1)
-	logrus.Info("Monitoring kubernetes for minecraft services")
-	go controller.Run(w.stop)
+	logrus.Info("Monitoring Kubernetes for Minecraft services")
+	go serviceController.Run(w.stop)
+	go statefulSetController.Run(w.stop)
```

Comment:
I'm not sure this is correct, I'm still learning about goroutines.

Reply:
This is right; however, I should have used `close(w.stop)` in the `Stop` function instead of the following to allow for multiple "listeners":

```
		w.stop <- struct{}{}
```

Reply:
Done now. (BTW Note that these statements moved around a bit, due to the conditional flag.)